File: packages/web-client/src/query-guard.ts
/**
 * The raw-query guard. `client.query()` and the React
 * `useRawSql` hook are the untrusted raw-SQL tier: an app hands us a SQL
 * string and we run it against the local database. Two rules make that safe
 * to expose, enforced HERE in the core (previously they lived in the
 * now-removed `@syncular/kysely` read-only driver):
 *
 *   1. READ-ONLY. Only `select / with / explain / pragma / values` are
 *      allowed. A write (`insert/update/delete/…`) against the local mirror
 *      bypasses the outbox (SPEC §7.1) and silently diverges from the
 *      server — writes MUST go through `client.mutate([...])`.
 *   2. ONE STATEMENT. `sqlite-wasm`'s `exec` runs every statement in a
 *      multi-statement string (`SELECT 1; DROP TABLE t`), while the native
 *      SQLite adapters prepare only the first. We unify on the strict
 *      behaviour: exactly one statement per `query()`.
 *
 * The guard only fronts the PUBLIC `client.query()` — engine-internal reads
 * call the `ClientDatabase` directly and are trusted, so they are never
 * routed through here.
 *
 * RFC 0005 adds a third rule: the previous-version context container is
 * engine-private local data and MUST NOT be raw-readable through the public
 * read tier. The check is engine-authoritative (SQLite's own `EXPLAIN` plan
 * plus `sqlite_master` root pages), never a regex over table names, because a
 * covering-index read opens the INDEX's root page rather than the table's and
 * quoted, schema-qualified, CTE, subquery and view forms all defeat text
 * matching.
 */
import type { ClientDatabase, SqlRow } from './database';

/**
 * RFC 0005 D3: the one non-reserved container table holding captured rows from
 * the previous local schema. Named here (not in the capture module) so the
 * guard stays dependency-free.
 */
export const PREVIOUS_VERSION_CONTAINER = 'syncular_prev_context';

/** Verbs a read-only query may begin with (lowercased). */
const READ_ONLY_VERBS = new Set([
  'select',
  'with',
  'explain',
  'pragma',
  'values',
]);

/** Raised when `client.query()` is handed SQL it will not run. */
export class RawSqlError extends Error {
  override readonly name = 'RawSqlError';
}

/**
 * Split `sql` into top-level statements at unquoted `;`, skipping over
 * string literals ('…'), quoted/bracketed identifiers ("…", `…`, […]) and
 * comments (-- …, /* … *​/) so a `;` inside any of them is not a boundary.
 * Returns the non-empty statements (comment/whitespace-only trailers drop).
 */
function splitStatements(sql: string): string[] {
  const statements: string[] = [];
  let start = 0;
  let i = 0;
  const n = sql.length;

  const pushIfNonEmpty = (end: number) => {
    const stripped = stripLeading(sql.slice(start, end));
    if (stripped.length > 0) statements.push(sql.slice(start, end));
    start = end + 1;
  };

  while (i < n) {
    const c = sql[i];
    if (c === '-' && sql[i + 1] === '-') {
      const nl = sql.indexOf('\n', i + 2);
      i = nl === -1 ? n : nl + 1;
    } else if (c === '/' && sql[i + 1] === '*') {
      const close = sql.indexOf('*/', i + 2);
      i = close === -1 ? n : close + 2;
    } else if (c === "'" || c === '"' || c === '`') {
      i = skipQuoted(sql, i, c as "'" | '"' | '`');
    } else if (c === '[') {
      const close = sql.indexOf(']', i + 1);
      i = close === -1 ? n : close + 1;
    } else if (c === ';') {
      pushIfNonEmpty(i);
      i += 1;
    } else {
      i += 1;
    }
  }
  pushIfNonEmpty(n);
  return statements;
}

/** Advance past a quoted run opened at `open`; SQL doubles the quote to escape it. */
function skipQuoted(sql: string, open: number, quote: "'" | '"' | '`'): number {
  let i = open + 1;
  const n = sql.length;
  while (i < n) {
    if (sql[i] === quote) {
      if (sql[i + 1] === quote) i += 2;
      else return i + 1;
    } else {
      i += 1;
    }
  }
  return n;
}

/** Strip leading whitespace and comments, returning the remainder. */
function stripLeading(sql: string): string {
  return sql
    .replace(/^\s*(?:--[^\n]*(?:\n|$)|\/\*[\s\S]*?\*\/|\s)+/, '')
    .trimStart();
}

function firstWords(sql: string): string {
  const trimmed = sql.trim().replace(/\s+/g, ' ');
  return trimmed.length > 72 ? `${trimmed.slice(0, 72)}…` : trimmed;
}

/**
 * The main verb of a `WITH …` statement: SQLite allows a with-clause before
 * SELECT **and before INSERT/UPDATE/DELETE**, so `WITH t AS (…) DELETE …`
 * must not slip through the verb allowlist. CTE bodies live inside
 * parentheses, and a bare keyword cannot be a CTE name, so the first
 * paren-depth-0 keyword after the clause IS the main verb.
 */
function mainVerbAfterWith(sql: string): string | undefined {
  const MAIN_VERBS = new Set([
    'select',
    'values',
    'insert',
    'update',
    'delete',
    'replace',
  ]);
  let depth = 0;
  let i = 0;
  const n = sql.length;
  let sawWith = false;
  while (i < n) {
    const c = sql[i] as string;
    if (c === '-' && sql[i + 1] === '-') {
      const nl = sql.indexOf('\n', i + 2);
      i = nl === -1 ? n : nl + 1;
    } else if (c === '/' && sql[i + 1] === '*') {
      const close = sql.indexOf('*/', i + 2);
      i = close === -1 ? n : close + 2;
    } else if (c === "'" || c === '"' || c === '`') {
      i = skipQuoted(sql, i, c as "'" | '"' | '`');
    } else if (c === '[') {
      const close = sql.indexOf(']', i + 1);
      i = close === -1 ? n : close + 1;
    } else if (c === '(') {
      depth += 1;
      i += 1;
    } else if (c === ')') {
      depth -= 1;
      i += 1;
    } else if (/[A-Za-z_]/.test(c)) {
      let j = i + 1;
      while (j < n && /[A-Za-z0-9_]/.test(sql[j] as string)) j += 1;
      const word = sql.slice(i, j).toLowerCase();
      if (depth === 0) {
        if (!sawWith && word === 'with') sawWith = true;
        else if (sawWith && MAIN_VERBS.has(word)) return word;
      }
      i = j;
    } else {
      i += 1;
    }
  }
  return undefined;
}

/**
 * Assert `sql` is a single read-only statement, or throw `RawSqlError`.
 * Called by `client.query()` before the string reaches the database.
 */
export function assertReadOnlyQuery(sql: string): void {
  const statements = splitStatements(sql);
  if (statements.length === 0) {
    throw new RawSqlError('client.query() was given an empty statement.');
  }
  if (statements.length > 1) {
    throw new RawSqlError(
      `client.query() runs a single statement, but ${statements.length} were ` +
        'given. Split them into separate query() calls. ' +
        `First: ${firstWords(statements[0] ?? '')}`,
    );
  }
  const statement = stripLeading(statements[0] ?? '');
  const verb = statement.match(/^([a-zA-Z]+)/)?.[1]?.toLowerCase();
  const rejectWrite = (): never => {
    throw new RawSqlError(
      'client.query() is read-only — this statement writes the local ' +
        'database directly, which bypasses the sync outbox (SPEC §7.1). Use ' +
        '`client.mutate([...])` for inserts/updates/deletes. ' +
        `Rejected: ${firstWords(sql)}`,
    );
  };
  if (verb === undefined || !READ_ONLY_VERBS.has(verb)) rejectWrite();
  if (verb === 'with') {
    // SQLite allows `WITH … DELETE/INSERT/UPDATE`; only a SELECT/VALUES
    // main statement is a read.
    const main = mainVerbAfterWith(statement);
    if (main !== 'select' && main !== 'values') rejectWrite();
  }
}

/**
 * RFC 0005 D4: the root pages that belong to the container table and every
 * index SQLite created on it (the `PRIMARY KEY (tbl, row_id)` autoindex).
 * `undefined` means the container does not exist, so no read can reach it.
 * Root pages are per-database, so this set is only meaningful against opcodes
 * with `p3 = 0` (`main`).
 */
function protectedRootPages(db: ClientDatabase): Set<number> | undefined {
  const rows = db.query(
    'SELECT rootpage FROM sqlite_master WHERE tbl_name = ?',
    [PREVIOUS_VERSION_CONTAINER],
  );
  if (rows.length === 0) return undefined;
  const pages = new Set<number>();
  for (const row of rows) {
    const page = Number(row.rootpage);
    if (Number.isSafeInteger(page) && page > 0) pages.add(page);
  }
  return pages;
}

/** True iff the RFC 0005 container currently exists locally. */
export function previousVersionContainerExists(db: ClientDatabase): boolean {
  return protectedRootPages(db) !== undefined;
}

/**
 * RFC 0005 D4: refuse a statement that reads `syncular_prev_context`.
 *
 * The caller MUST run this inside the same transaction that executes `sql` on
 * the same connection: the protected root-page set is re-read on every call
 * and never cached, so a schema change cannot land between the decision and
 * the read. The check fails closed while the container exists: an unavailable
 * or unparseable `EXPLAIN`, or any `OpenRead`/`OpenWrite` against an attached
 * or `temp` database, is a refusal.
 */
export function assertProtectedTableAccess(
  db: ClientDatabase,
  sql: string,
): void {
  const pages = protectedRootPages(db);
  if (pages === undefined) return;
  const refuse = (detail: string): never => {
    throw new RawSqlError(
      `client.query() refused a statement that reads the ${PREVIOUS_VERSION_CONTAINER} ` +
        `container (RFC 0005): ${detail}. Read captured previous-version rows ` +
        'only through client.previousVersionSnapshot().',
    );
  };
  let plan: SqlRow[];
  try {
    plan = db.query(`EXPLAIN ${sql}`);
  } catch {
    return refuse('EXPLAIN is unavailable');
  }
  if (plan.length === 0) return refuse('EXPLAIN returned no plan');
  let parsed = false;
  for (const row of plan) {
    const opcode = row.opcode;
    if (typeof opcode !== 'string') continue;
    parsed = true;
    if (opcode !== 'OpenRead' && opcode !== 'OpenWrite') continue;
    if (Number(row.p3) !== 0) {
      return refuse('the statement opens an attached or temporary database');
    }
    if (pages.has(Number(row.p2))) {
      return refuse('the statement opens the protected table or its index');
    }
  }
  if (!parsed) return refuse('the EXPLAIN plan could not be parsed');
}

File: rust/crates/client/src/query_guard.rs
//! The raw-query guard (DESIGN-queries.md I3), mirrored from the TS core's
//! `query-guard.ts` — the two cores enforce the same rule so `query` behaves
//! identically through every binding:
//!
//! 1. READ-ONLY: only `select / with / explain / pragma / values` may run.
//!    A write against the local mirror bypasses the outbox (SPEC §7.1) and
//!    silently diverges from the server — writes go through `mutate`.
//! 2. ONE STATEMENT: `sqlite3_prepare` parses only the first statement and
//!    silently ignores the tail; instead of ignoring, reject loudly.
//!
//! Engine-internal reads use the connection directly and never route here.

const READ_ONLY_VERBS: [&str; 5] = ["select", "with", "explain", "pragma", "values"];
const PUBLIC_SYNC_VERSION: &str = "_sync_version";
const NATIVE_SYNC_VERSION: &str = "_syncular_version";

/// Lower the app-facing row-version pseudo-column to the Rust client's private
/// physical column. The TypeScript client stores `_sync_version` directly;
/// Rust keeps `_syncular_version` for SSP2 SQLite-image parity, but both query
/// surfaces must accept the same authored SQL.
///
/// Only SQL identifiers are rewritten. Single-quoted values and comments are
/// copied byte-for-byte so application data cannot be changed by lowering.
pub(crate) fn lower_public_query_sql(sql: &str) -> String {
    let bytes = sql.as_bytes();
    let mut output = String::with_capacity(sql.len());
    let mut copied_through = 0usize;
    let mut i = 0usize;
    while i < bytes.len() {
        match bytes[i] {
            b'-' if bytes.get(i + 1) == Some(&b'-') => {
                i = sql[i..]
                    .find('\n')
                    .map_or(bytes.len(), |offset| i + offset + 1);
            }
            b'/' if bytes.get(i + 1) == Some(&b'*') => {
                i = sql[i + 2..]
                    .find("*/")
                    .map_or(bytes.len(), |offset| i + 2 + offset + 2);
            }
            b'\'' => {
                i += 1;
                while i < bytes.len() {
                    if bytes[i] == b'\'' {
                        if bytes.get(i + 1) == Some(&b'\'') {
                            i += 2;
                        } else {
                            i += 1;
                            break;
                        }
                    } else {
                        i += 1;
                    }
                }
            }
            quote @ (b'"' | b'`') => {
                let content_start = i + 1;
                let mut end = content_start;
                while end < bytes.len() && bytes[end] != quote {
                    end += 1;
                }
                if end < bytes.len()
                    && sql[content_start..end].eq_ignore_ascii_case(PUBLIC_SYNC_VERSION)
                {
                    output.push_str(&sql[copied_through..content_start]);
                    output.push_str(NATIVE_SYNC_VERSION);
                    copied_through = end;
                }
                i = (end + 1).min(bytes.len());
            }
            b'[' => {
                let content_start = i + 1;
                let end = sql[content_start..]
                    .find(']')
                    .map_or(bytes.len(), |offset| content_start + offset);
                if end < bytes.len()
                    && sql[content_start..end].eq_ignore_ascii_case(PUBLIC_SYNC_VERSION)
                {
                    output.push_str(&sql[copied_through..content_start]);
                    output.push_str(NATIVE_SYNC_VERSION);
                    copied_through = end;
                }
                i = (end + 1).min(bytes.len());
            }
            byte if byte.is_ascii_alphabetic() || byte == b'_' => {
                let start = i;
                i += 1;
                while i < bytes.len() && (bytes[i].is_ascii_alphanumeric() || bytes[i] == b'_') {
                    i += 1;
                }
                if sql[start..i].eq_ignore_ascii_case(PUBLIC_SYNC_VERSION) {
                    output.push_str(&sql[copied_through..start]);
                    output.push_str(NATIVE_SYNC_VERSION);
                    copied_through = i;
                }
            }
            _ => i += 1,
        }
    }
    output.push_str(&sql[copied_through..]);
    output
}

/// Assert `sql` is one read-only statement; `Err(message)` otherwise.
pub fn assert_read_only_query(sql: &str) -> Result<(), String> {
    let statements = split_statements(sql);
    if statements.is_empty() {
        return Err("query() was given an empty statement".to_owned());
    }
    if statements.len() > 1 {
        return Err(format!(
            "query() runs a single statement, but {} were given — split them \
             into separate query() calls",
            statements.len()
        ));
    }
    let head = strip_leading(statements[0]);
    let verb: String = head
        .chars()
        .take_while(|c| c.is_ascii_alphabetic())
        .collect::<String>()
        .to_ascii_lowercase();
    let reject = || {
        Err(format!(
            "query() is read-only — this statement writes the local database \
             directly, which bypasses the sync outbox (SPEC §7.1); use \
             mutate() for inserts/updates/deletes. Rejected: {}",
            first_words(sql)
        ))
    };
    if !READ_ONLY_VERBS.contains(&verb.as_str()) {
        return reject();
    }
    if verb == "with" {
        // SQLite allows `WITH … DELETE/INSERT/UPDATE`; only a SELECT/VALUES
        // main statement is a read.
        match main_verb_after_with(head) {
            Some(main) if main == "select" || main == "values" => {}
            _ => return reject(),
        }
    }
    Ok(())
}

/// The main verb of a `WITH …` statement. CTE bodies live inside
/// parentheses, and a bare keyword cannot be a CTE name, so the first
/// paren-depth-0 keyword after the clause is the main verb.
fn main_verb_after_with(sql: &str) -> Option<String> {
    const MAIN_VERBS: [&str; 6] = ["select", "values", "insert", "update", "delete", "replace"];
    let bytes = sql.as_bytes();
    let n = bytes.len();
    let mut depth: i32 = 0;
    let mut i = 0usize;
    let mut saw_with = false;
    while i < n {
        match bytes[i] {
            b'-' if bytes.get(i + 1) == Some(&b'-') => {
                i = sql[i..].find('\n').map_or(n, |off| i + off + 1);
            }
            b'/' if bytes.get(i + 1) == Some(&b'*') => {
                i = sql[i + 2..].find("*/").map_or(n, |off| i + 2 + off + 2);
            }
            q @ (b'\'' | b'"' | b'`') => {
                i += 1;
                while i < n {
                    if bytes[i] == q {
                        if bytes.get(i + 1) == Some(&q) {
                            i += 2;
                        } else {
                            i += 1;
                            break;
                        }
                    } else {
                        i += 1;
                    }
                }
            }
            b'[' => {
                i = sql[i + 1..].find(']').map_or(n, |off| i + 1 + off + 1);
            }
            b'(' => {
                depth += 1;
                i += 1;
            }
            b')' => {
                depth -= 1;
                i += 1;
            }
            c if c.is_ascii_alphabetic() || c == b'_' => {
                let mut j = i + 1;
                while j < n && (bytes[j].is_ascii_alphanumeric() || bytes[j] == b'_') {
                    j += 1;
                }
                let word = sql[i..j].to_ascii_lowercase();
                if depth == 0 {
                    if !saw_with && word == "with" {
                        saw_with = true;
                    } else if saw_with && MAIN_VERBS.contains(&word.as_str()) {
                        return Some(word);
                    }
                }
                i = j;
            }
            _ => i += 1,
        }
    }
    None
}

/// Split into top-level statements at unquoted `;`, skipping string
/// literals, quoted/bracketed identifiers and comments. Statements that are
/// only whitespace/comments are dropped.
fn split_statements(sql: &str) -> Vec<&str> {
    let bytes = sql.as_bytes();
    let mut statements = Vec::new();
    let mut start = 0usize;
    let mut i = 0usize;
    let n = bytes.len();
    while i < n {
        match bytes[i] {
            b'-' if bytes.get(i + 1) == Some(&b'-') => {
                i = sql[i..].find('\n').map_or(n, |off| i + off + 1);
            }
            b'/' if bytes.get(i + 1) == Some(&b'*') => {
                i = sql[i + 2..].find("*/").map_or(n, |off| i + 2 + off + 2);
            }
            q @ (b'\'' | b'"' | b'`') => {
                i += 1;
                while i < n {
                    if bytes[i] == q {
                        if bytes.get(i + 1) == Some(&q) {
                            i += 2; // doubled quote escapes itself
                        } else {
                            i += 1;
                            break;
                        }
                    } else {
                        i += 1;
                    }
                }
            }
            b'[' => {
                i = sql[i + 1..].find(']').map_or(n, |off| i + 1 + off + 1);
            }
            b';' => {
                let piece = &sql[start..i];
                if !strip_leading(piece).is_empty() {
                    statements.push(piece);
                }
                start = i + 1;
                i += 1;
            }
            _ => i += 1,
        }
    }
    let piece = &sql[start..];
    if !strip_leading(piece).is_empty() {
        statements.push(piece);
    }
    statements
}

/// Strip leading whitespace and comments.
fn strip_leading(sql: &str) -> &str {
    let mut rest = sql.trim_start();
    loop {
        if let Some(after) = rest.strip_prefix("--") {
            rest = after.find('\n').map_or("", |off| &after[off + 1..]);
        } else if let Some(after) = rest.strip_prefix("/*") {
            rest = after.find("*/").map_or("", |off| &after[off + 2..]);
        } else {
            return rest;
        }
        rest = rest.trim_start();
    }
}

fn first_words(sql: &str) -> String {
    let compact = sql.split_whitespace().collect::<Vec<_>>().join(" ");
    if compact.chars().count() > 72 {
        let head: String = compact.chars().take(72).collect();
        format!("{head}…")
    } else {
        compact
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn allows_read_only_verbs() {
        for sql in [
            "SELECT 1",
            "  select * from tasks",
            "WITH x AS (SELECT 1) SELECT * FROM x",
            "WITH x AS (SELECT 1), y AS (SELECT 2) SELECT * FROM x, y",
            "WITH RECURSIVE c(n) AS (VALUES (1)) SELECT n FROM c",
            "EXPLAIN QUERY PLAN SELECT 1",
            "PRAGMA table_info(tasks)",
            "VALUES (1), (2)",
            "-- leading comment\nSELECT 1",
            "/* block */ SELECT 1",
            "SELECT 1;",
            "SELECT 1; -- done",
            "SELECT ';' AS s",
            "SELECT 'it''s; fine'",
            "SELECT \"a;b\" FROM tasks",
            "SELECT [a;b] FROM tasks",
        ] {
            assert!(assert_read_only_query(sql).is_ok(), "should allow: {sql}");
        }
    }

    #[test]
    fn rejects_writes_and_multi_statements() {
        for sql in [
            "INSERT INTO tasks (id) VALUES ('t1')",
            "UPDATE tasks SET title = 'x'",
            "DELETE FROM tasks",
            "DROP TABLE tasks",
            "CREATE TABLE evil (id)",
            "BEGIN",
            "VACUUM",
            "SELECT 1; DROP TABLE tasks",
            "SELECT 1; SELECT 2",
            "WITH t AS (SELECT 1) DELETE FROM tasks",
            "WITH t AS (SELECT 1) INSERT INTO tasks (id) SELECT 'x'",
            "WITH t AS (SELECT 1) UPDATE tasks SET title = 'x'",
            "",
            "   ",
            "-- only a comment",
            ";",
        ] {
            assert!(assert_read_only_query(sql).is_err(), "should reject: {sql}");
        }
    }

    #[test]
    fn lowers_only_app_facing_sync_version_identifiers() {
        let sql = "SELECT t._sync_version AS server_version, \"_sync_version\" AS quoted \
                   FROM tasks t WHERE note = '_sync_version' /* _sync_version */ \
                   -- _sync_version\nORDER BY t.id";
        let lowered = lower_public_query_sql(sql);
        assert!(lowered.contains("t._syncular_version AS server_version"));
        assert!(lowered.contains("\"_syncular_version\" AS quoted"));
        assert!(lowered.contains("note = '_sync_version'"));
        assert!(lowered.contains("/* _sync_version */"));
        assert!(lowered.contains("-- _sync_version"));
    }
}
